cmd/soroban-rpc/internal/methods/simulate_transaction.go

```diff
@@ -49,45 +130,85 @@ func NewSimulateTransactionHandler(logger *log.Entry, coreClient *stellarcore.Cl
 				Error: "Transaction contains more than one operation",
 			}
 		}
+		op := txEnvelope.Operations()[0]
 
-		var sourceAccount string
-		if opSourceAccount := txEnvelope.Operations()[0].SourceAccount; opSourceAccount != nil {
-			sourceAccount = opSourceAccount.ToAccountId().Address()
+		var sourceAccount xdr.AccountId
+		if opSourceAccount := op.SourceAccount; opSourceAccount != nil {
+			sourceAccount = opSourceAccount.ToAccountId()
 		} else {
-			sourceAccount = txEnvelope.SourceAccount().ToAccountId().Address()
+			// FIXME: SourceAccount() panics, so, the user can doctor an envelope which makes the server crash
+			sourceAccount = txEnvelope.SourceAccount().ToAccountId()
 		}
 
-		xdrOp, ok := txEnvelope.Operations()[0].Body.GetInvokeHostFunctionOp()
+		xdrOp, ok := op.Body.GetInvokeHostFunctionOp()
 		if !ok {
 			return SimulateTransactionResponse{
 				Error: "Transaction does not contain invoke host function operation",
 			}
 		}
 
-		coreResponse, err := coreClient.Preflight(ctx, sourceAccount, xdrOp)
+		hfB64, err := xdr.MarshalBase64(xdrOp.Function)
 		if err != nil {
-			logger.WithError(err).WithField("request", request).
-				Info("could not submit preflight request to core")
 			return SimulateTransactionResponse{
-				Error: "Could not submit request to core",
+				Error: "Cannot marshal host function",
 			}
 		}
+		hfCString := C.CString(hfB64)
+		sourceAccountB64, err := xdr.MarshalBase64(sourceAccount)
+		if err != nil {
+			return SimulateTransactionResponse{
+				Error: "Cannot marshal source account",
+			}
+		}
+		readTx, err := db.NewLedgerEntryReaderTx()
+		if err != nil {
+			return SimulateTransactionResponse{
+				Error: "Cannot create db transaction",
+			}
+		}
+		defer readTx.Done()
+		latestLedger, err := readTx.GetLatestLedgerSequence()
+		if err != nil {
+			return SimulateTransactionResponse{
+				Error: "Cannot read latest ledger",
+			}
+		}
+		li := C.CLedgerInfo{
+			network_passphrase: C.CString(networkPassphrase),
+			sequence_number:    C.uint(latestLedger),
+			protocol_version:   20,
+			timestamp:          C.uint64_t(time.Now().Unix()),
+			// Current base reserve is 0.5XLM (in stroops)
+			base_reserve: 5_000_000,
+		}
+
+		handle := C.uintptr_t(cgo.NewHandle(snapshotSourceHandle{readTx, logger}))
```

Comment:
I guess this doesn't need a matching free?

Reply:
You are absolutely right